File: python/aurora_examples.py
#!/usr/bin/env python3
import argparse, json, os, sys, urllib.parse
import requests

ENV = os.getenv("ENV", "sandbox")
BASE = f"https://{ENV}.tflapis.com"
TOKEN = os.getenv("TOKEN", "<TOKEN>")
HEADERS = {
    "Authorization": f"Bearer {TOKEN}",
    "Accept": "text/plain, application/json",
}

def _post(path: str, body: dict):
    resp = requests.post(BASE + path, headers={**HEADERS, "Content-Type": "application/json"}, json=body)
    if not resp.ok:
        print(f"ERROR {resp.status_code}: {resp.text}", file=sys.stderr)
        resp.raise_for_status()
    return resp.json()

def _get(path: str):
    resp = requests.get(BASE + path, headers=HEADERS)
    if not resp.ok:
        print(f"ERROR {resp.status_code}: {resp.text}", file=sys.stderr)
        resp.raise_for_status()
    return resp.json()

def build_customer(args):
    customer = {
        "firstName": args.first_name or "First",
        "lastName": args.last_name or "Last",
        "email": args.email or "dev@example.com",
    }
    if args.phone: customer["phoneNumber"] = args.phone
    a1 = args.address1
    if any([a1, args.city, args.region, args.postal, args.country]):
        missing = [n for n,v in [("address1",a1),("city",args.city),("region",args.region),("postalCode",args.postal),("country",args.country)] if not v]
        if missing:
            raise SystemExit(f"If any address field is set, the following are required (address2 optional): {', '.join(missing)}")
        customer["address"] = {
            "address1": a1,
            "address2": args.address2,
            "city": args.city,
            "region": args.region,
            "postalCode": args.postal,
            "country": args.country,
        }
    return customer

def query_events(args):
    q = urllib.parse.quote(args.search)
    print(json.dumps(_get(f"/Catalog/Events?query={q}&perPage=10&page=1"), indent=2))

def query_tickets(args):
    print(json.dumps(_get(f"/Catalog/Events/{args.event_id}/Tickets"), indent=2))

def query_autocomplete(args):
    q = urllib.parse.quote(args.search)
    print(json.dumps(_get(f"/Catalog/Autocomplete?searchText={q}&catalogs=event&catalogs=performer&catalogs=venue&catalogs=category"), indent=2))

def managed_checkout(args):
    cart = _post("/Cart", {"items": []})
    cart_id = cart.get("id")
    if not cart_id: raise SystemExit("No cart id returned")

    _post(f"/Cart/{cart_id}/Items", {
        "listingId": args.listing_id,
        "quantity": args.quantity,
        "currencyType": args.currency,
        "price": float(args.price),
    })

    body = {"customer": build_customer(args)}
    if args.client_order_id:
        body["clientOrderIdentifier"] = args.client_order_id

    print(json.dumps(_post(f"/Cart/{cart_id}/Checkout", body), indent=2))

def unmanaged_checkout(args):
    body = {
        "customer": build_customer(args),
        "shoppingCart": {
            "items": [{
                "listingId": args.listing_id,
                "quantity": args.quantity,
                "currencyType": args.currency,
                "price": float(args.price),
            }]
        }
    }
    if args.client_order_id:
        body["clientOrderIdentifier"] = args.client_order_id
    print(json.dumps(_post("/Cart/Checkout", body), indent=2))

def common_checkout_args(sp):
    sp.add_argument("--email", default=os.getenv("EMAIL"))
    sp.add_argument("--first-name", default=os.getenv("FIRST_NAME"))
    sp.add_argument("--last-name", default=os.getenv("LAST_NAME"))
    sp.add_argument("--phone", default=os.getenv("PHONE"))
    sp.add_argument("--address1", default=os.getenv("ADDRESS1"))
    sp.add_argument("--address2", default=os.getenv("ADDRESS2"))
    sp.add_argument("--city", default=os.getenv("CITY"))
    sp.add_argument("--region", default=os.getenv("REGION"))
    sp.add_argument("--postal", default=os.getenv("POSTAL"))
    sp.add_argument("--country", default=os.getenv("COUNTRY"))
    sp.add_argument("--client-order-id")

parser = argparse.ArgumentParser(description="Aurora API Examples — Python")
sub = parser.add_subparsers(dest="cmd", required=True)

sp = sub.add_parser("query-events")
sp.add_argument("search")
sp.set_defaults(func=query_events)

sp = sub.add_parser("query-tickets")
sp.add_argument("event_id")
sp.set_defaults(func=query_tickets)

sp = sub.add_parser("query-autocomplete")
sp.add_argument("search")
sp.set_defaults(func=query_autocomplete)

sp = sub.add_parser("managed-checkout")
sp.add_argument("listing_id")
sp.add_argument("quantity", type=int)
sp.add_argument("price", type=float)
sp.add_argument("currency")
common_checkout_args(sp)
sp.set_defaults(func=managed_checkout)

sp = sub.add_parser("unmanaged-checkout")
sp.add_argument("listing_id")
sp.add_argument("quantity", type=int)
sp.add_argument("price", type=float)
sp.add_argument("currency")
common_checkout_args(sp)
sp.set_defaults(func=unmanaged_checkout)

if __name__ == "__main__":
  if TOKEN == "<TOKEN>":
    sys.exit("Set TOKEN env var (export TOKEN=<TOKEN>)")
  args = parser.parse_args()
  args.func(args)

File: javascript/aurora_examples.js
#!/usr/bin/env node
/**
 * Aurora API Examples — JavaScript (Node 18+)
 *
 * Usage:
 *   TOKEN=<TOKEN> ENV=sandbox node aurora_examples.js query-events "Chiefs"
 *   TOKEN=<TOKEN> ENV=sandbox node aurora_examples.js query-tickets <EVENT_ID>
 *   TOKEN=<TOKEN> ENV=sandbox node aurora_examples.js query-autocomplete "Taylor Swift"
 *
 *   TOKEN=<TOKEN> ENV=sandbox node aurora_examples.js managed-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [--email ... --first ... --last ... --phone ... --address1 ... --address2 ... --city ... --region ... --postal ... --country ... --client-order-id ...]
 *
 *   TOKEN=<TOKEN> ENV=sandbox node aurora_examples.js unmanaged-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [same customer flags]
 */

const TOKEN = process.env.TOKEN || "<TOKEN>";
const ENV = process.env.ENV || "sandbox";
const BASE = `https://${ENV}.tflapis.com`;

if (TOKEN === "<TOKEN>") {
  console.error("Set TOKEN env var (e.g., TOKEN=<TOKEN>)");
  process.exit(1);
}

// ---------- tiny arg parser ----------
const args = process.argv.slice(2);
const cmd = args[0];
const positional = [];
const flags = {};
for (let i = 1; i < args.length; i++) {
  const a = args[i];
  if (a.startsWith("--")) {
    const key = a.replace(/^--/, "");
    const val = args[i + 1] && !args[i + 1].startsWith("--") ? args[++i] : "true";
    flags[key] = val;
  } else {
    positional.push(a);
  }
}

async function get(path) {
  const res = await fetch(`${BASE}${path}`, {
    headers: {
      Authorization: `Bearer ${TOKEN}`,
      Accept: "text/plain, application/json",
    },
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`GET ${path} -> ${res.status}: ${text}`);
  }
  return res.json();
}

async function post(path, body) {
  const res = await fetch(`${BASE}${path}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${TOKEN}`,
      Accept: "text/plain, application/json",
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body ?? {}),
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`POST ${path} -> ${res.status}: ${text}`);
  }
  return res.json();
}

// ---------- customer builder with address rule ----------
function buildCustomer(f) {
  const customer = {
    firstName: f.first ?? process.env.FIRST_NAME ?? "First",
    lastName: f.last ?? process.env.LAST_NAME ?? "Last",
    email: f.email ?? process.env.EMAIL ?? "dev@example.com",
  };
  if (f.phone ?? process.env.PHONE) customer.phoneNumber = f.phone ?? process.env.PHONE;

  const address1 = f.address1 ?? process.env.ADDRESS1;
  const address2 = f.address2 ?? process.env.ADDRESS2 ?? "";
  const city = f.city ?? process.env.CITY;
  const region = f.region ?? process.env.REGION;
  const postal = f.postal ?? process.env.POSTAL;
  const country = f.country ?? process.env.COUNTRY;

  const any = [address1, city, region, postal, country].some(Boolean);
  if (any) {
    const missing = [];
    if (!address1) missing.push("address1");
    if (!city) missing.push("city");
    if (!region) missing.push("region");
    if (!postal) missing.push("postal");
    if (!country) missing.push("country");
    if (missing.length) {
      throw new Error(
        `If any address field is set, the following are required (address2 optional): ${missing.join(", ")}`
      );
    }
    customer.address = { address1, address2, city, region, postalCode: postal, country };
  }
  return customer;
}

// ---------- commands ----------
async function queryEvents(search) {
  const q = encodeURIComponent(search);
  const data = await get(`/Catalog/Events?query=${q}&perPage=10&page=1`);
  console.log(JSON.stringify(data, null, 2));
}

async function queryTickets(eventId) {
  const data = await get(`/Catalog/Events/${eventId}/Tickets`);
  console.log(JSON.stringify(data, null, 2));
}

async function queryAutocomplete(search) {
  const q = encodeURIComponent(search);
  const data = await get(
    `/Catalog/Autocomplete?searchText=${q}&catalogs=event&catalogs=performer&catalogs=venue&catalogs=category`
  );
  console.log(JSON.stringify(data, null, 2));
}

async function managedCheckout(listingId, qty, price, currency) {
  // 1) Create cart
  const cart = await post("/Cart", { items: [] });
  const cartId = cart.id;
  if (!cartId) throw new Error("No cart id returned from /Cart");

  // 2) Add item (single-object body)
  await post(`/Cart/${cartId}/Items`, {
    listingId,
    quantity: Number(qty),
    currencyType: currency,
    price: Number(price),
  });

  // 3) Checkout w/ customer (clientOrderIdentifier optional)
  const customer = buildCustomer({
    email: flags.email,
    first: flags.first,
    last: flags.last,
    phone: flags.phone,
    address1: flags.address1,
    address2: flags.address2,
    city: flags.city,
    region: flags.region,
    postal: flags.postal,
    country: flags.country,
  });

  const body = { customer };
  if (flags["client-order-id"]) body.clientOrderIdentifier = flags["client-order-id"];

  const out = await post(`/Cart/${cartId}/Checkout`, body);
  console.log(JSON.stringify(out, null, 2));
}

async function unmanagedCheckout(listingId, qty, price, currency) {
  const customer = buildCustomer({
    email: flags.email,
    first: flags.first,
    last: flags.last,
    phone: flags.phone,
    address1: flags.address1,
    address2: flags.address2,
    city: flags.city,
    region: flags.region,
    postal: flags.postal,
    country: flags.country,
  });

  const body = {
    customer,
    shoppingCart: {
      items: [
        {
          listingId,
          quantity: Number(qty),
          currencyType: currency,
          price: Number(price),
        },
      ],
    },
  };
  if (flags["client-order-id"]) body.clientOrderIdentifier = flags["client-order-id"];

  const out = await post("/Cart/Checkout", body);
  console.log(JSON.stringify(out, null, 2));
}

// ---------- main ----------
(async () => {
  try {
    switch (cmd) {
      case "query-events":
        if (!positional[0]) throw new Error("Usage: query-events <SEARCH>");
        await queryEvents(positional[0]);
        break;
      case "query-tickets":
        if (!positional[0]) throw new Error("Usage: query-tickets <EVENT_ID>");
        await queryTickets(positional[0]);
        break;
      case "query-autocomplete":
        if (!positional[0]) throw new Error("Usage: query-autocomplete <SEARCH>");
        await queryAutocomplete(positional[0]);
        break;
      case "managed-checkout":
        if (positional.length < 4) throw new Error("Usage: managed-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [--email ... --first ... --last ... --phone ... --address1 ... --address2 ... --city ... --region ... --postal ... --country ... --client-order-id ...]");
        await managedCheckout(positional[0], positional[1], positional[2], positional[3]);
        break;
      case "unmanaged-checkout":
        if (positional.length < 4) throw new Error("Usage: unmanaged-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [--email ... --first ... --last ... ...]");
        await unmanagedCheckout(positional[0], positional[1], positional[2], positional[3]);
        break;
      case undefined:
      case "help":
      case "--help":
      case "-h":
        console.log(`Aurora API Examples — JavaScript

Env:
  TOKEN=<TOKEN> (required)
  ENV=sandbox|prod (default: sandbox)

Commands:
  query-events <SEARCH>
  query-tickets <EVENT_ID>
  query-autocomplete <SEARCH>
  managed-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [--email ... --first ... --last ... --phone ... --address1 ... --address2 ... --city ... --region ... --postal ... --country ... --client-order-id ...]
  unmanaged-checkout <LISTING_ID> <QTY> <PRICE> <CURRENCY> [same flags]

Examples:
  TOKEN=<TOKEN> node aurora_examples.js query-events "Chiefs"
  TOKEN=<TOKEN> node aurora_examples.js query-tickets 1972578
  TOKEN=<TOKEN> node aurora_examples.js query-autocomplete "Taylor Swift"

  TOKEN=<TOKEN> node aurora_examples.js managed-checkout <LISTING_ID> 2 26.00 USD --email dev@example.com --first Jane --last Doe --phone "555-555-1234" --address1 "1313 Mockingbird Lane" --city "Kansas City" --region "MO" --postal "64106" --country "US"

  TOKEN=<TOKEN> node aurora_examples.js unmanaged-checkout <LISTING_ID> 2 26.00 USD --email dev@example.com --first Jane --last Doe
`);
        break;
      default:
        throw new Error(`Unknown command: ${cmd}`);
    }
  } catch (err) {
    console.error(String(err?.message || err));
    process.exit(1);
  }
})();